File: 700-y.rb
Y = ->(z) {
  U.(
    ->(f) {
      recursion = ->(n) { f.(f).(n) }
      z.(recursion)
    }
  )
}

fact = Y.(
  ->(recursion) {
    ->(x) {
      if x == 1
        1
      else
        x * recursion.(x-1)
      end
    }
  }
)

puts fact.(5)

#Refactor - inline recursion and inline U

File: js/700-y.js
const U = (f) => { return f(f) }
const Z = (z) =>{
  return U((f) => {
    const recur = (n) => { return f(f)(n) }
    return z(recur)
  })
}

const fact = Z((recur) => {
  return (x) => {
    if (x === 0) {
      return 1
    } else {
      return x * recur(x-1)
    }
  }
})

console.log(fact(8))

//Refactor - inline recursion and inline U
